File: python/src/savings_account.py
from src.account import Account


class SavingsAccount(Account):
    def __init__(self, account_holder, balance, account_number, interest_rate):
        super().__init__(account_holder, balance, account_number)
        # TODO: Implement constructor
        self.interest_rate = interest_rate

    def get_interest_rate(self):
        # TODO: Implement getter
        return self.interest_rate

    def set_interest_rate(self, interest_rate):
        # TODO: Implement setter
        if interest_rate < 0:
            print("Invalid Interest rate.")
        else:
            self.interest_rate = interest_rate

    def apply_interest(self):
        # TODO: Implement method to apply interest to the balance
        # New balance = current balance + (current balance * interest rate)
        self.balance += (self.balance * self.interest_rate)

    def debit(self, amount):
        # TODO: Implement debit method
        # Savings accounts have overdraft protection - don't allow balance to go negative
        if self.balance - amount >=0:
            self.balance -= amount

File: src/main/java/io/zipcoder/SavingsAccount.java
package io.zipcoder;

public class SavingsAccount extends Account {
    private Double interestRate;

    public SavingsAccount(Object accountHolder, Double balance, String accountNumber, Double interestRate) {
        super(accountHolder, balance, accountNumber);
        // TODO: Implement constructor
        this.interestRate = interestRate;
    }

    public Double getInterestRate() {
        // TODO: Implement getter
        return interestRate;
    }

    public void setInterestRate(Double interestRate) {
        // TODO: Implement setter
        if(interestRate > 0) {
            this.interestRate = interestRate;
        } else {
            System.out.println("Invalid Input.");
        }
    }

    public void applyInterest() {
        // TODO: Implement method to apply interest to the balance
        // New balance = current balance + (current balance * interest rate)
        double interestAmount = getBalance() * interestRate;
        setBalance(getBalance() + interestAmount);
    }

    @Override
    public void debit(Double amount) {
        // TODO: Implement debit method
        // Savings accounts have overdraft protection - don't allow balance to go negative
        if(amount > getBalance()){
            System.out.println("Transaction declined.");
        } else {
            super.debit(amount);
        }
    }
}
